File: sdk/go/agent_client.go
package stratium

import (
	"context"
	"fmt"
	"time"

	agpb "github.com/stratiumdata/go-sdk/gen/services/agent-gateway"
	"google.golang.org/grpc"
)

// AgentConfig holds configuration for agent authentication.
type AgentConfig struct {
	AgentID            string            // Registered agent ID
	AgentSecret        string            // Agent client secret
	GatewayAddress     string            // Agent Gateway address (e.g., "localhost:50054")
	DefaultPurpose     string            // Default delegation purpose
	DefaultMaxTier     ActionTier        // Default max action tier
	ClassificationCaps map[string]string // Default per-hierarchy caps
}

// ActionTier mirrors the protobuf ActionTier enum for SDK consumers.
type ActionTier int32

const (
	ActionTierReasoning      ActionTier = 0
	ActionTierReadOnly       ActionTier = 1
	ActionTierInternalModify ActionTier = 2
	ActionTierExternalComms  ActionTier = 3
	ActionTierDestructive    ActionTier = 4
)

// Delegation holds the result of a delegation creation.
type Delegation struct {
	Token            string
	DelegationID     string
	Depth            int
	ExpiresAt        time.Time
	RootDelegationID string
}

// AgentAction describes an action to execute through the Agent Gateway.
type AgentAction struct {
	TargetService      string            // "platform", "key-manager", "key-access"
	Method             string            // gRPC method path
	Action             string            // "read", "write", "delete", etc.
	ActionTier         ActionTier        // Sensitivity tier
	ToolName           string            // Tool being invoked
	ResourceAttributes map[string]string // Resource metadata for policy evaluation
	Payload            []byte            // Serialized proto request
}

// ActionResult holds the result of an agent action execution.
type ActionResult struct {
	Authorized bool
	Payload    []byte // Serialized proto response (if authorized)
	Error      string
	Decision   *CompoundDecisionInfo
}

// CompoundDecisionInfo provides the decision breakdown for debugging.
type CompoundDecisionInfo struct {
	DeniedAtDepth   int
	DeniedPrincipal string
}

// DelegationOptions configures a new delegation.
type DelegationOptions struct {
	ApprovedTools       []string
	ApprovedActions     []ActionTier
	MaxActionTier       ActionTier
	ClassificationCaps  map[string]string
	ResourceConstraints map[string]string
	Purpose             string
	TTLSeconds          int
	ConversationID      string
}

// PlannedAgentAction describes an action for validation.
type PlannedAgentAction struct {
	ToolName     string
	Action       string
	DeclaredTier ActionTier
	Resource     map[string]string
	Parameters   map[string]string
}

// ValidationResult holds the result of action plan validation.
type ValidationResult struct {
	Valid   bool
	Results []ActionValidation
}

// ActionValidation holds the result of a single action validation.
type ActionValidation struct {
	ActionIndex  int
	Valid        bool
	Reason       string
	ActualTier   ActionTier
	TierMismatch bool
}

// AgentClient wraps a standard Stratium client with agent delegation capabilities.
type AgentClient struct {
	*Client

	agentConfig    *AgentConfig
	gatewayConn    *grpc.ClientConn
	gatewayClient  agpb.AgentGatewayServiceClient
	activeDelegation *Delegation
}

// NewAgentClient creates a new agent-aware Stratium client.
func NewAgentClient(userConfig *Config, agentConfig *AgentConfig) (*AgentClient, error) {
	if agentConfig == nil {
		return nil, fmt.Errorf("agent config cannot be nil")
	}
	if agentConfig.AgentID == "" {
		return nil, fmt.Errorf("agent_id is required")
	}
	if agentConfig.GatewayAddress == "" {
		return nil, fmt.Errorf("gateway_address is required")
	}

	// Create the base client
	baseClient, err := NewClient(userConfig)
	if err != nil {
		return nil, fmt.Errorf("failed to create base client: %w", err)
	}

	// Connect to the Agent Gateway
	dialOpts := baseClient.config.dialOptions()
	gatewayConn, err := grpc.NewClient(agentConfig.GatewayAddress, dialOpts...)
	if err != nil {
		baseClient.Close()
		return nil, fmt.Errorf("failed to connect to agent gateway: %w", err)
	}

	return &AgentClient{
		Client:        baseClient,
		agentConfig:   agentConfig,
		gatewayConn:   gatewayConn,
		gatewayClient: agpb.NewAgentGatewayServiceClient(gatewayConn),
	}, nil
}

// CreateDelegation creates a delegation for this user+agent pair.
func (c *AgentClient) CreateDelegation(ctx context.Context, opts *DelegationOptions) (*Delegation, error) {
	if opts == nil {
		opts = &DelegationOptions{}
	}

	// Apply defaults from agent config
	if opts.Purpose == "" {
		opts.Purpose = c.agentConfig.DefaultPurpose
	}
	if opts.MaxActionTier == 0 && c.agentConfig.DefaultMaxTier > 0 {
		opts.MaxActionTier = c.agentConfig.DefaultMaxTier
	}
	if opts.ClassificationCaps == nil && c.agentConfig.ClassificationCaps != nil {
		opts.ClassificationCaps = c.agentConfig.ClassificationCaps
	}
	if opts.TTLSeconds == 0 {
		opts.TTLSeconds = 900 // 15 min default
	}

	// Convert action tiers to proto enum
	protoActions := make([]agpb.ActionTier, len(opts.ApprovedActions))
	for i, a := range opts.ApprovedActions {
		protoActions[i] = agpb.ActionTier(a)
	}

	ctx, err := c.attachAuth(ctx)
	if err != nil {
		return nil, fmt.Errorf("authentication failed: %w", err)
	}

	resp, err := c.gatewayClient.CreateDelegation(ctx, &agpb.CreateDelegationRequest{
		AgentId:             c.agentConfig.AgentID,
		ApprovedTools:       opts.ApprovedTools,
		ApprovedActions:     protoActions,
		MaxActionTier:       agpb.ActionTier(opts.MaxActionTier),
		ClassificationCaps:  opts.ClassificationCaps,
		ResourceConstraints: opts.ResourceConstraints,
		Purpose:             opts.Purpose,
		TtlSeconds:          int32(opts.TTLSeconds),
		ConversationId:      opts.ConversationID,
	})
	if err != nil {
		return nil, fmt.Errorf("create delegation failed: %w", err)
	}

	delegation := &Delegation{
		Token:            resp.DelegationToken,
		DelegationID:     resp.DelegationId,
		Depth:            int(resp.Depth),
		RootDelegationID: resp.RootDelegationId,
	}
	if resp.ExpiresAt != nil {
		delegation.ExpiresAt = resp.ExpiresAt.AsTime()
	}

	c.activeDelegation = delegation
	return delegation, nil
}

// ExecuteAction executes an action through the Agent Gateway with double-hop auth.
func (c *AgentClient) ExecuteAction(ctx context.Context, action *AgentAction) (*ActionResult, error) {
	if c.activeDelegation == nil {
		return nil, fmt.Errorf("no active delegation — call CreateDelegation first")
	}

	resp, err := c.gatewayClient.ExecuteAction(ctx, &agpb.ExecuteActionRequest{
		DelegationToken:    c.activeDelegation.Token,
		TargetService:      action.TargetService,
		Method:             action.Method,
		Action:             action.Action,
		ActionTier:         agpb.ActionTier(action.ActionTier),
		ToolName:           action.ToolName,
		ResourceAttributes: action.ResourceAttributes,
		Payload:            action.Payload,
	})
	if err != nil {
		return nil, fmt.Errorf("execute action failed: %w", err)
	}

	result := &ActionResult{
		Authorized: resp.Authorized,
		Payload:    resp.Payload,
		Error:      resp.Error,
	}
	if resp.Decision != nil {
		result.Decision = &CompoundDecisionInfo{
			DeniedAtDepth:   int(resp.Decision.DeniedAtDepth),
			DeniedPrincipal: resp.Decision.DeniedPrincipal,
		}
	}

	return result, nil
}

// ValidateActionPlan validates a set of planned actions before execution.
func (c *AgentClient) ValidateActionPlan(ctx context.Context, actions []PlannedAgentAction) (*ValidationResult, error) {
	if c.activeDelegation == nil {
		return nil, fmt.Errorf("no active delegation — call CreateDelegation first")
	}

	protoActions := make([]*agpb.PlannedAction, len(actions))
	for i, a := range actions {
		protoActions[i] = &agpb.PlannedAction{
			ToolName:     a.ToolName,
			Action:       a.Action,
			DeclaredTier: agpb.ActionTier(a.DeclaredTier),
			Resource:     a.Resource,
			Parameters:   a.Parameters,
		}
	}

	resp, err := c.gatewayClient.ValidateActionPlan(ctx, &agpb.ValidateActionPlanRequest{
		DelegationId: c.activeDelegation.DelegationID,
		Actions:      protoActions,
	})
	if err != nil {
		return nil, fmt.Errorf("validate action plan failed: %w", err)
	}

	result := &ValidationResult{Valid: resp.Valid}
	for _, r := range resp.Results {
		result.Results = append(result.Results, ActionValidation{
			ActionIndex:  int(r.ActionIndex),
			Valid:        r.Valid,
			Reason:       r.Reason,
			ActualTier:   ActionTier(r.ActualTier),
			TierMismatch: r.TierMismatch,
		})
	}

	return result, nil
}

// RevokeDelegation revokes the active delegation.
func (c *AgentClient) RevokeDelegation(ctx context.Context, reason string) error {
	if c.activeDelegation == nil {
		return fmt.Errorf("no active delegation")
	}

	_, err := c.gatewayClient.RevokeDelegation(ctx, &agpb.RevokeDelegationRequest{
		DelegationId: c.activeDelegation.DelegationID,
		Reason:       reason,
	})
	if err != nil {
		return fmt.Errorf("revoke delegation failed: %w", err)
	}

	c.activeDelegation = nil
	return nil
}

// GetActiveDelegation returns the current active delegation, or nil.
func (c *AgentClient) GetActiveDelegation() *Delegation {
	return c.activeDelegation
}

// Close closes the agent client and all connections.
func (c *AgentClient) Close() error {
	if c.gatewayConn != nil {
		c.gatewayConn.Close()
	}
	return c.Client.Close()
}

// attachAuth adds the OIDC bearer token to the gRPC context.
func (c *AgentClient) attachAuth(ctx context.Context) (context.Context, error) {
	if c.auth == nil {
		return ctx, nil
	}
	token, err := c.auth.GetToken(ctx)
	if err != nil {
		return ctx, err
	}
	return contextWithAuth(ctx, token), nil
}

File: sdk/python/stratium_sdk/agent_client.py
"""
Agent-aware Stratium client for AI agent authorization with delegation tokens.

Provides double-hop authorization where every agent action is evaluated against
both the user's permissions and the agent's permissions within a delegation context.

Example::

    from stratium_sdk import StratiumConfig, AgentConfig, StratiumAgentClient, ActionTier

    config = StratiumConfig(
        platform_address="localhost:50051",
        key_access_address="localhost:50053",
        bearer_token="<user-oidc-token>",
    )
    agent_config = AgentConfig(
        agent_id="<registered-agent-id>",
        agent_secret="<agent-client-secret>",
        gateway_address="localhost:50054",
    )

    client = StratiumAgentClient(config, agent_config)
    client.initialize()

    delegation = client.create_delegation(
        purpose="Research documents",
        max_action_tier=ActionTier.READ_ONLY,
        approved_tools=["read_file", "search"],
        classification_caps={"nato": "CONFIDENTIAL"},
    )

    result = client.execute_action(
        action="read",
        tool_name="read_file",
        action_tier=ActionTier.READ_ONLY,
        target_service="key-access",
        method="/key_access.KeyAccessService/UnwrapDEK",
        resource_attributes={"classification": "CONFIDENTIAL", "hierarchy": "nato"},
    )
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import IntEnum
from typing import Dict, List, Optional, Sequence

import grpc

from .auth import TokenProvider, build_token_provider
from .config import StratiumConfig
from .errors import APIError, ValidationError
from .grpc_clients.base import create_channel


class ActionTier(IntEnum):
    """Sensitivity level of an action."""

    REASONING = 0
    READ_ONLY = 1
    INTERNAL_MODIFY = 2
    EXTERNAL_COMMS = 3
    DESTRUCTIVE = 4


class AgentTrustTier(IntEnum):
    """Trust level assigned to an agent."""

    UNVERIFIED = 0
    REGISTERED = 1
    CERTIFIED = 2
    PLATFORM_TRUSTED = 3


@dataclass
class AgentConfig:
    """Configuration for agent authentication."""

    agent_id: str
    agent_secret: str
    gateway_address: str
    default_purpose: str = ""
    default_max_tier: ActionTier = ActionTier.READ_ONLY
    classification_caps: Dict[str, str] = field(default_factory=dict)

    def validate(self) -> None:
        if not self.agent_id:
            raise ValidationError("agent_id is required")
        if not self.gateway_address:
            raise ValidationError("gateway_address is required")


@dataclass
class Delegation:
    """Represents an active delegation token."""

    token: str
    delegation_id: str
    depth: int
    expires_at: datetime
    root_delegation_id: str


@dataclass
class AgentAction:
    """An action to execute through the Agent Gateway."""

    action: str
    tool_name: str
    action_tier: ActionTier = ActionTier.READ_ONLY
    target_service: str = ""
    method: str = ""
    resource_attributes: Dict[str, str] = field(default_factory=dict)
    payload: bytes = b""


@dataclass
class ActionResult:
    """Result of an agent action execution."""

    authorized: bool
    payload: bytes = b""
    error: str = ""
    denied_at_depth: int = -1
    denied_principal: str = ""


@dataclass
class PlannedAction:
    """An action to validate before execution."""

    tool_name: str
    action: str
    declared_tier: ActionTier = ActionTier.READ_ONLY
    resource: Dict[str, str] = field(default_factory=dict)
    parameters: Dict[str, str] = field(default_factory=dict)


@dataclass
class ActionValidationItem:
    """Validation result for a single action."""

    action_index: int
    valid: bool
    reason: str = ""
    actual_tier: ActionTier = ActionTier.REASONING
    tier_mismatch: bool = False


@dataclass
class ValidationResult:
    """Result of action plan validation."""

    valid: bool
    results: List[ActionValidationItem] = field(default_factory=list)


class StratiumAgentClient:
    """
    Agent-aware Stratium client with delegation token flows.

    Wraps a standard StratiumConfig with agent credentials and provides
    methods for creating delegations, executing actions through the gateway,
    and validating action plans.
    """

    def __init__(
        self,
        user_config: StratiumConfig,
        agent_config: AgentConfig,
        token_provider: Optional[TokenProvider] = None,
    ) -> None:
        self._user_config = user_config
        self._agent_config = agent_config
        self._agent_config.validate()

        self._token_provider = token_provider or build_token_provider(user_config)
        self._gateway_channel: Optional[grpc.Channel] = None
        self._gateway_stub = None
        self._active_delegation: Optional[Delegation] = None
        self._initialized = False

    def initialize(self) -> None:
        """Establish connection to the Agent Gateway."""
        if self._initialized:
            return

        self._gateway_channel = create_channel(
            self._agent_config.gateway_address,
            self._user_config.use_tls,
        )
        # Import the generated stub lazily to avoid hard dependency on proto gen
        try:
            from .proto.services.agent_gateway import (
                agent_gateway_pb2_grpc as ag_grpc,
            )

            self._gateway_stub = ag_grpc.AgentGatewayServiceStub(self._gateway_channel)
        except ImportError:
            raise ImportError(
                "Agent Gateway proto stubs not found. "
                "Generate them with: cd sdk/python && make generate-proto"
            )

        self._initialized = True

    def create_delegation(
        self,
        purpose: str = "",
        max_action_tier: ActionTier = ActionTier.READ_ONLY,
        approved_tools: Optional[Sequence[str]] = None,
        approved_actions: Optional[Sequence[ActionTier]] = None,
        classification_caps: Optional[Dict[str, str]] = None,
        resource_constraints: Optional[Dict[str, str]] = None,
        ttl_seconds: int = 900,
        conversation_id: str = "",
    ) -> Delegation:
        """Create a delegation for this user+agent pair."""
        self._ensure_initialized()
        from .proto.services.agent_gateway import agent_gateway_pb2 as ag_pb

        purpose = purpose or self._agent_config.default_purpose
        if classification_caps is None:
            classification_caps = self._agent_config.classification_caps

        proto_actions = [ag_pb.ActionTier.Value(f"ACTION_TIER_{a.name}") for a in (approved_actions or [])]

        request = ag_pb.CreateDelegationRequest(
            agent_id=self._agent_config.agent_id,
            approved_tools=list(approved_tools or []),
            approved_actions=proto_actions,
            max_action_tier=int(max_action_tier),
            classification_caps=classification_caps or {},
            resource_constraints=resource_constraints or {},
            purpose=purpose,
            ttl_seconds=ttl_seconds,
            conversation_id=conversation_id,
        )

        metadata = self._auth_metadata()
        response = self._gateway_stub.CreateDelegation(request, metadata=metadata)

        expires_at = datetime.fromtimestamp(
            response.expires_at.seconds, tz=timezone.utc
        ) if response.expires_at else datetime.now(timezone.utc)

        delegation = Delegation(
            token=response.delegation_token,
            delegation_id=response.delegation_id,
            depth=response.depth,
            expires_at=expires_at,
            root_delegation_id=response.root_delegation_id,
        )
        self._active_delegation = delegation
        return delegation

    def execute_action(
        self,
        action: str,
        tool_name: str,
        action_tier: ActionTier = ActionTier.READ_ONLY,
        target_service: str = "",
        method: str = "",
        resource_attributes: Optional[Dict[str, str]] = None,
        payload: bytes = b"",
    ) -> ActionResult:
        """Execute an action through the Agent Gateway with double-hop auth."""
        self._ensure_initialized()
        self._ensure_delegation()
        from .proto.services.agent_gateway import agent_gateway_pb2 as ag_pb

        request = ag_pb.ExecuteActionRequest(
            delegation_token=self._active_delegation.token,
            target_service=target_service,
            method=method,
            action=action,
            action_tier=int(action_tier),
            tool_name=tool_name,
            resource_attributes=resource_attributes or {},
            payload=payload,
        )

        response = self._gateway_stub.ExecuteAction(request)

        result = ActionResult(
            authorized=response.authorized,
            payload=response.payload,
            error=response.error,
        )
        if response.decision:
            result.denied_at_depth = response.decision.denied_at_depth
            result.denied_principal = response.decision.denied_principal
        return result

    def validate_action_plan(
        self,
        actions: Sequence[PlannedAction],
    ) -> ValidationResult:
        """Validate a set of planned actions before execution."""
        self._ensure_initialized()
        self._ensure_delegation()
        from .proto.services.agent_gateway import agent_gateway_pb2 as ag_pb

        proto_actions = []
        for a in actions:
            proto_actions.append(
                ag_pb.PlannedAction(
                    tool_name=a.tool_name,
                    action=a.action,
                    declared_tier=int(a.declared_tier),
                    resource=a.resource,
                    parameters=a.parameters,
                )
            )

        request = ag_pb.ValidateActionPlanRequest(
            delegation_id=self._active_delegation.delegation_id,
            actions=proto_actions,
        )

        response = self._gateway_stub.ValidateActionPlan(request)

        return ValidationResult(
            valid=response.valid,
            results=[
                ActionValidationItem(
                    action_index=r.action_index,
                    valid=r.valid,
                    reason=r.reason,
                    actual_tier=ActionTier(r.actual_tier),
                    tier_mismatch=r.tier_mismatch,
                )
                for r in response.results
            ],
        )

    def revoke_delegation(self, reason: str = "") -> None:
        """Revoke the active delegation."""
        self._ensure_initialized()
        self._ensure_delegation()
        from .proto.services.agent_gateway import agent_gateway_pb2 as ag_pb

        request = ag_pb.RevokeDelegationRequest(
            delegation_id=self._active_delegation.delegation_id,
            reason=reason,
        )
        self._gateway_stub.RevokeDelegation(request)
        self._active_delegation = None

    @property
    def active_delegation(self) -> Optional[Delegation]:
        """Return the current active delegation, or None."""
        return self._active_delegation

    def close(self) -> None:
        """Close the gateway channel."""
        if self._gateway_channel is not None:
            self._gateway_channel.close()
            self._gateway_channel = None
        self._initialized = False

    def __enter__(self) -> StratiumAgentClient:
        self.initialize()
        return self

    def __exit__(self, *args) -> None:
        self.close()

    # --- Internal ---

    def _ensure_initialized(self) -> None:
        if not self._initialized:
            raise APIError("client not initialized — call initialize() first")

    def _ensure_delegation(self) -> None:
        if self._active_delegation is None:
            raise APIError("no active delegation — call create_delegation() first")

    def _auth_metadata(self) -> List[tuple]:
        """Build gRPC metadata with the bearer token."""
        if self._token_provider is None:
            return []
        token = self._token_provider.get_token()
        if token:
            return [("authorization", f"Bearer {token}")]
        return []
